Mathlib/RingTheory/PowerSeries/Eval.lean

```diff
@@ -0,0 +1,345 @@
+    (p.map f).EvalEq (f x) (f a) := by
+  intro i
+  rw [trunc_map, Polynomial.eval_map, Polynomial.eval₂_hom, ← map_sub, ← map_pow]
+  exact map_dvd f (h i)
+
+lemma evalEq_map_comp_C (f : R⟦X⟧ →+* S) (p : R⟦X⟧) :
+    (p.map (f.comp (C R))).EvalEq (f .X) (f p) :=
+  (evalEq_map_C_X p).map f
+
+lemma evalEq_map_algebraMap [Algebra R S] (f : R⟦X⟧ →ₐ[R] S) (p : R⟦X⟧) :
+    (p.map (algebraMap R S)).EvalEq (f .X) (f p) :=
+  f.comp_algebraMap ▸ evalEq_map_comp_C f.toRingHom p
+
+@[ext (iff := false) 1100]
+lemma ringHom_ext (f g : R⟦X⟧ →+* S)
+    (hC : f.comp (C R) = g.comp (C R))
+    (hX : f .X = g .X) (hX' : IsHausdorffAt (f .X) := by infer_instance) :
+    f = g :=
+  RingHom.ext fun p ↦
+    EvalEq.ext (evalEq_map_comp_C f p) (Eq.symm hC ▸ Eq.symm hX ▸ evalEq_map_comp_C g p)
+
+@[ext (iff := false) 1100]
+lemma algHom_ext [Algebra R S] (f g : R⟦X⟧ →ₐ[R] S)
+    (hX : f .X = g .X) (hX' : IsHausdorffAt (f .X) := by infer_instance) :
+    f = g :=
+  AlgHom.ext (DFunLike.congr_fun (ringHom_ext f.toRingHom g.toRingHom
+    (f.comp_algebraMap.trans g.comp_algebraMap.symm) hX hX') : _)
+
+variable (x) in
+/-- `convergesAt x` is the set of power series that converge at `x`. -/
+def convergesAt : Subalgebra (Polynomial R) R⟦X⟧ where
+  carrier := { p | ∃ a, p.EvalEq x a }
+  mul_mem' {p q} hp hq := by
+    obtain ⟨⟨a, ha⟩, ⟨b, hb⟩⟩ := And.intro hp hq
+    exact ⟨a * b, ha.mul hb⟩
+  add_mem' {p q} hp hq := by
+    obtain ⟨⟨a, ha⟩, ⟨b, hb⟩⟩ := And.intro hp hq
+    exact ⟨a + b, ha.add hb⟩
+  algebraMap_mem' p := ⟨p.eval x, evalEq_coe x p⟩
+
+lemma convergesAt_eq_top_iff :
+    convergesAt x = ⊤ ↔ IsPrecompleteAt x := by
+  simp only [IsPrecompleteAt, isPrecomplete_iff, Ideal.span_singleton_pow, smul_eq_mul,
+    Ideal.mul_top, SModEq.sub_mem, Ideal.mem_span_singleton, ← top_le_iff, SetLike.le_def,
+    Algebra.mem_top, forall_const]
+  constructor
+  · intro H f hf
+    choose g hg using fun i : ℕ ↦ dvd_sub_comm.mp (hf i.le_succ)
+    simp_rw [sub_eq_iff_eq_add] at hg
+    let p := mk g + C R (f 0)
+    obtain ⟨L, hL⟩ := @H p
+    have (n) : (trunc (n + 1) p).eval x = f (n + 1) := by
+      induction n with
+      | zero => simp [p, hg]
+      | succ n IH =>
+        rw [hg, trunc_succ, Polynomial.eval_add, IH, add_comm]
+        simp [p, mul_comm]
+    refine ⟨L, fun n ↦ ?_⟩
+    cases n
+    · simp
+    rw [← this]
+    exact hL _
+  · intro hx p
+    refine hx (fun i ↦ (p.trunc i).eval x) (fun {m n} h ↦ ?_)
+    obtain ⟨r, hr⟩ : .X ^ m ∣ trunc m p - trunc n p := by
+      rw [← trunc_trunc_of_le _ h]
+      simp +contextual [Polynomial.X_pow_dvd_iff, coeff_trunc _ m]
+    rw [← Polynomial.eval_sub, hr, mul_comm, Polynomial.eval_mul_X_pow]
+    exact dvd_mul_left _ _
+
+lemma convergesAt_eq_top (x : R) [IsPrecompleteAt x] :
+    convergesAt x = ⊤ := convergesAt_eq_top_iff.mpr ‹_›
+
+/-- `p.eval x` is the evaluation of `p : R⟦X⟧` at `x`, and `0` if the former is not well defined. -/
+noncomputable
+def eval (p : R⟦X⟧) (x : R) : R :=
```

Comment:
Under what circumstances does this coincide with the appropriate topological sum `tsum`?